DAT-15338 Refactor connection string patterns to allow extensions to extend they

vitaliimak opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
<!--  Mandatory Labels to add to a PR

At least one of the labels must be added to the PR before it's merged. If no label is provided the workflow will fail and you will not be able to merge the PR. After the label is added it re-runs the `Pull Request Labels / label (pull_request)` and gives a green check. 

`skipReleaseNotes`   - Don't show up on the Draft Release Notes page
`notableChanges`     - Any notable changes
`TypeEnhancement`    - New features
`TypeTest`           - New Test features
`TypeBug`            - bug fixes
`breakingChanges`    - any breaking changes
`APIBreakingChanges` - any API breaking changes
`sdou`               - Security, Driver and Other Updates -dependabot PR's
`newContributors`    - New Contributors -->


## Description

<!--
A clear and concise description of the change being made.  

- Introduce what was/will be done in the title
  - Titles show in release notes and search results, so make them useful
  - Use verbs at the beginning of the title, such as "fix", "implement", "improve", "update", and "add" 
  - Be specific about what was fixed or changed
  - Good Example: `Fix the --should-snapshot-data CLI parameter to be preserved when the --data-output-directory property is not specified in the command.`
  - Bad Example: `Fixed --should-snapshot-data`  
- If there is an existing issue this addresses, include "Fixes #XXXX" to auto-link the issue to this PR
- If there is NOT an existing issue, consider creating one.
  - In general, issues describe wanted change from an end-user perspective and PRs describe the technical change.
  - If this change is very small and not worth splitting off an issue, include `Steps To Reproduce`, `Expected Behavior`, and `Actual Behavior` sections in this PR as you would have in the issue.
- Describe what users need and how the fix will affect them
- Describe how the code change addresses the problem
- Ensure private information is redacted.
-->

## Things to be aware of

<!--
- Describe the technical choices you made
- Describe impacts on the codebase
-->

## Things to worry about

<!--
- List any questions or concerns you have with the change
- List unknowns you have 
-->

## Additional Context

<!--
Add any other context about the problem here.
-->

StevenMassaro reviewed (approved):
This is outside the scope of this work, but this class would do with a refactor to use the service locator. This would allow plugins to provide their own database-specific filter patterns without requiring changes here.

vitaliimak commented:
Hi @StevenMassaro, if it is possible I'd like to do it. It looks like a much better approach than hardcoding new extension patterns inside the core codebase. The current fix cannot affect extension before the next liquibase release, so we do have time to make refactoring.
Do you agree @filipelautert?

vitaliimak commented:
Hello @StevenMassaro, @filipelautert 
Could you look at this variant of solving the issue?

StevenMassaro reviewed (approved):
Looks great, just needs some javadoc. This implementation will go a long way in improving the extensibility of Liquibase. Nice job!

filipelautert reviewed (approved):
After adding javadocs as Steve commented this will be fine!

vitaliimak commented:
Hello @filipelautert @StevenMassaro 
I've made more changes than only adding Javadoc. 
I've replaced the ConnectionPatterns interface with the abstract class, as it should store patterns. The abstract class is a better place for holding data than an interface. It allowed me to decrease fields visibility and add "setter" methods.
I've added Javadoc only for "setter" methods (public void addJdbcBlankPatterns(Map.Entry<Pattern, Pattern> jdbcBlankPatterns)), as API clients should use them to add more patterns. The getter methods are used only in the `JdbcConnection` class in the liquibase-core, I don't know if I need to add Javadoc to them as well.
What do you think?

StevenMassaro commented on liquibase-standard/src/main/java/liquibase/database/jvm/JdbcConnection.java:
```diff
@@ -609,7 +580,11 @@ public boolean supportsBatchUpdates() throws DatabaseException {
     }
 
-    private static class PatternPair {
+    /**
+     *  @deprecated this class has been moved to {@link ConnectionPatterns}, use that one instead.
+     */
+    @Deprecated
+    public static class PatternPair {
```
Can this class be deleted? If it was private before, it will not be a breaking change.

vitaliimak replied:
Oh! I missed that it was private. I'll remove it. Thanks!

StevenMassaro reviewed (approved):
Looks good!
